engine/crates/composition/src/compose/object.rs

```diff
@@ -24,3 +26,67 @@ pub(super) fn merge_field_arguments<'a>(
         .map(|arg| (arg.argument_name().id, arg.argument_type().id))
         .collect()
 }
+
+pub(super) fn compose_object_fields<'a>(first: FieldWalker<'a>, fields: &[FieldWalker<'a>], ctx: &mut Context<'a>) {
+    if fields
+        .iter()
+        .filter(|f| !(f.is_shareable() || f.is_external() || f.is_key()))
+        .count()
+        > 1
+    {
+        let next = &fields[1];
+
+        ctx.diagnostics.push_fatal(format!(
+            "The field `{}` on `{}` is defined in two subgraphs (`{}` and `{}`).",
+            first.name().as_str(),
+            first.parent_definition().name().as_str(),
+            first.parent_definition().subgraph().name().as_str(),
+            next.parent_definition().subgraph().name().as_str(),
+        ));
+    }
+
+    let first_is_key = first.is_key();
+    if fields.iter().any(|field| field.is_key() != first_is_key) {
+        let name = format!(
+            "{}.{}",
+            first.parent_definition().name().as_str(),
+            first.name().as_str()
+        );
+        let (key_subgraphs, non_key_subgraphs) = fields
+            .iter()
+            .partition::<Vec<FieldWalker<'_>>, _>(|field| field.is_key());
+
+        ctx.diagnostics.push_fatal(format!(
+            "The field `{name}` is part of `@key` in {} but not in {}",
+            key_subgraphs
+                .into_iter()
+                .map(|f| f.parent_definition().subgraph().name().as_str())
+                .join(", "),
+            non_key_subgraphs
+                .into_iter()
+                .map(|f| f.parent_definition().subgraph().name().as_str())
+                .join(", "),
+        ));
+    }
```

Comment:
I kind of thought subgraphs could have different sets of keys, so not sure about this.  Though I'm not sure if I've actually read that somewhere or if it was just an assumption on my part

Reply:
They can: https://www.apollographql.com/docs/federation/entities-advanced#differing-keys-across-subgraphs